File: done/dmopc14c2p5.py
import sys
input = sys.stdin.readline
n, m = map(int, input().split())
graph = [[] for i in range(n)]
prob = [0.0 for i in range(n)]
for i in range(m):
    a, b = map(int, input().split())
    graph[a - 1].append(b - 1)
prob[0] = 1.0
for i in range(n):
    if len(graph[i]) != 0:
        temp = prob[i] / len(graph[i])
        for neighbour in graph[i]:
            prob[neighbour] += temp
    else:
        print(prob[i])

File: done/dmopc14c2p5.cpp
#include <bits/stdc++.h>
using namespace std;

int a, b, x, y;
vector<int> graph[1000000];
double prob[1000000];

int main() {
  scanf("%d %d", &a, &b);

  for (int i = 0; i < b; i++) {
    scanf("%d %d", &x, &y);
    graph[x - 1].push_back(y - 1);
  }

  prob[0] = 1;
  for (int i = 0; i < a; i++) {
    const double temp = prob[i] / graph[i].size();
    for (int j = 0; j < graph[i].size(); j++) {
      prob[graph[i][j]] += temp;
    }
    if (graph[i].size() == 0) {
      printf("%.12lf\n", prob[i]);
    }
  }
}